crates/red_knot/src/salsa_db/semantic/ast_ids.rs

```diff
@@ -0,0 +1,406 @@
+                    self.visit_annotation(expr);
+                }
+
+                self.deferred.push_back(body);
+            }
+
+            Stmt::ClassDef(StmtClassDef {
+                arguments,
+                body,
+                decorator_list,
+                type_params,
+                name: _,
+                range: _,
+            }) => {
+                for decorator in decorator_list {
+                    self.visit_decorator(decorator);
+                }
+
+                if let Some(type_params) = type_params {
+                    self.visit_type_params(type_params);
+                }
+
+                if let Some(arguments) = arguments {
+                    self.visit_arguments(arguments);
+                }
+
+                self.deferred.push_back(body);
+            }
+
+            _ => preorder::walk_stmt(self, stmt),
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        let expression_id = self
+            .expressions
+            .push(unsafe { AstNodeRef::new(self.parsed.clone(), expr) });
+
+        self.expressions_map
+            .insert(NodeKey::from_node(expr.into()), expression_id);
+
+        preorder::walk_expr(self, expr);
+    }
+}
+
+#[derive(Clone)]
+struct AstNodeRef<T> {
+    _parsed: Arc<Parsed<ModModule>>,
+    node: std::ptr::NonNull<T>,
+}
+
+impl<T> AstNodeRef<T> {
+    /// Creates a new `AstNodeRef` that reference `node`. The `parsed` is the parsed module to which
+    /// the `AstNodeRef` belongs.
+    ///
+    /// The function is marked as `unsafe` because it is the caller's responsibility to ensure that
+    /// `node` is part of `parsed`'s AST. Dereferencing the value if that's not the case is UB
+    /// because it can then point to a no longer valid memory location.
+    #[allow(unsafe_code)]
+    unsafe fn new(parsed: Arc<Parsed<ModModule>>, node: &T) -> Self {
+        Self {
+            _parsed: parsed,
+            node: std::ptr::NonNull::from(node),
+        }
+    }
+
+    pub fn node(&self) -> &T {
+        // SAFETY: Holding on to `parsed` ensures that the AST to which `node` belongs is still alive
+        // and not moved.
+        #[allow(unsafe_code)]
+        unsafe {
+            self.node.as_ref()
+        }
+    }
+}
+
+impl<T> Deref for AstNodeRef<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        self.node()
+    }
+}
+
+impl<T> std::fmt::Debug for AstNodeRef<T>
+where
+    T: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("AstNodeRef").field(&self.node()).finish()
+    }
+}
+
+impl<T> PartialEq for AstNodeRef<T>
+where
+    T: PartialEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.node() == other.node()
+    }
+}
+
+impl<T> Eq for AstNodeRef<T> where T: Eq {}
+
+#[allow(unsafe_code)]
+unsafe impl<T> Send for AstNodeRef<T> where T: Send {}
+#[allow(unsafe_code)]
+unsafe impl<T> Sync for AstNodeRef<T> where T: Sync {}
```

Comment:
Because `Send` and `Sync` are `unsafe` traits. Rust normally implements these traits automatically for you if all the fields of your struct are `Send` or `Sync`. However, Rust can't do that for raw pointers, because it don't know what the guarantees are about the pointer. 

In our case, `Send` and `Sync` is safe because we never mutate the pointer nor the data they're referencing (our data structure is similar to an `Arc`)